src/long.ts

```diff
@@ -1021,9 +1021,20 @@ export class Long {
     if (options && options.relaxed) return this.toNumber();
     return { $numberLong: this.toString() };
   }
-  static fromExtendedJSON(doc: { $numberLong: string }, options?: EJSONOptions): number | Long {
-    const result = Long.fromString(doc.$numberLong);
-    return options && options.relaxed ? result.toNumber() : result;
+  static fromExtendedJSON(
+    doc: { $numberLong: string },
+    options?: EJSONOptions
+  ): number | Long | bigint {
+    const longResult = Long.fromString(doc.$numberLong);
+    const defaults = { useBigInt64: false, relaxed: true };
+    const ejsonOptions = { ...defaults, ...options };
+    if (ejsonOptions.useBigInt64) {
+      return BigInt(doc.$numberLong);
```

Comment:
Unlike in BSON and the 'number' case for EJSON, there isn't restrictions on value size or length.

Luckily BigInt casting function will throw an error for bad syntax unlike Number (yay!) (can we add a test? `{$numberLong: "blah!"}`)

But we have added support here for syntax that should not work: 
- `{ $numberLong: "0xA" }` should fail
- `{ $numberLong: "0o7" }` should fail
- `{ $numberLong: "0b1" }` should fail

And we've unintentionally allowed strings of any length to be provided here, I'm not sure what would be the best approach here, we can either figure out some way to assert some maximum length. Or we can construct a Long.fromString and use long.toBigInt here, asserting that the fromString logic would prevent an extremely long string from being converted.

Reply:
From discussion: We want to assert a maximum length on the string, this isn't the same as what Long does but our newer API can be better about preventing malformed input. There should be no leading zeros permitted but a plus or minus sign needs to be supported. We can iterate the string or use regexp to assert this.

Consider making a new BSONError subclass

Reply:
> There should be no leading zeros permitted

They are currently permitted when deserializing to `Long`, as far as I can tell. I would recommend to keep validation consistent between deserializing to `Long` and deserializing to `bigint` (and changing validation for deserializing to `Long` would probably be breaking, but this might be exactly the right time for that if it’s something you want to do).

Reply:
If we're all in agreement (@dariakp @baileympearson @durran @W-A-James) I think we can do the validation for both `bigint` and `Long` cases, I was aiming to only impact the bigint case here since it would be opt-in, but useBigInt64 and increasing the validation are divergent concerns so we shouldn't overload the option's behavior.

We certainly don't want to add support for the hex, oct, and bin formats just because useBigInt64 is turned on, so that assertion I think we're all on board with.

`Long.fromString` currently allows a number of things that probably should be errors like 
* `"-----------123"`
* `"00000000123"`
* `"12.23"` - becomes 12
* `'8'.repeat(25)` - too large, can be determined by string length (gets truncated)
* `"9223372036854775808"` - too large int64.max + 1 (gets truncated)

We can decide to apply all, some, or none of the above validations, I think the most valuable and least pedantic is applying a string length check, but they're all great if we want to improve things :)

Reply:
I agree that at the very least, we should add the string length check to inputs of `fromExtendedJSON` in both the `bigInt` and `Long` cases to keep with the canonical EJSON spec and to prevent hanging on strings that would be too long to be a valid 64-bit integer and may be the result of malicious user input.

Canonical EJSON representation of a BSON.Long:

> {"$numberLong": <64-bit signed integer as a string>}

Reply:
Updating with results of Slack discussion;

Going to change bigint case to truncate rather than throw an error on encountering a bigint outside of int64 range to be consistent with behaviour of `Long.toString`